File: tools/python/reference/rustinfer_reference/environment.py
"""Fail-closed, standard-library-only primary-host environment binding.

The probe intentionally runs before importing or loading a model backend.  It
uses only immutable host facts exposed by ``nvidia-smi``, ``/proc`` and
``/etc/os-release``.  The BF16 fact is derived from the observed CUDA compute
capability; each BF16 model backend additionally checks the selected CUDA
device through its pinned runtime before executing a model.
"""

from __future__ import annotations

import math
import platform
import re
import shutil
import subprocess
from collections.abc import Callable, Sequence
from pathlib import Path

from .constants import (
    PRIMARY_CPU_LOGICAL_THREADS,
    PRIMARY_CPU_MODEL,
    PRIMARY_CPU_PHYSICAL_CORES,
    PRIMARY_CPU_GOVERNOR,
    PRIMARY_CPU_GOVERNOR_POLICY_COUNT,
    PRIMARY_DRIVER_CUDA_API_VERSION,
    PRIMARY_ENVIRONMENT_ID,
    PRIMARY_GPU_COMPUTE_CAPABILITY,
    PRIMARY_GPU_COUNT,
    PRIMARY_GPU_MEMORY_MIB,
    PRIMARY_GPU_NAME,
    PRIMARY_KERNEL_RELEASE,
    PRIMARY_MACHINE,
    PRIMARY_NVIDIA_DRIVER_VERSION,
    PRIMARY_OS_ID,
    PRIMARY_OS_VERSION_ID,
    PRIMARY_RAM_BYTES,
)

ENVIRONMENT_SNAPSHOT_SCHEMA_VERSION = "1.0.0"


class EnvironmentContractError(ValueError):
    """The observed host cannot claim the primary environment ID."""


PRIMARY_ENVIRONMENT_SNAPSHOT: dict[str, object] = {
    "schema_version": ENVIRONMENT_SNAPSHOT_SCHEMA_VERSION,
    "environment_id": PRIMARY_ENVIRONMENT_ID,
    "host": {
        "os_id": PRIMARY_OS_ID,
        "os_version_id": PRIMARY_OS_VERSION_ID,
        "kernel_release": PRIMARY_KERNEL_RELEASE,
        "machine": PRIMARY_MACHINE,
        "cpu_model": PRIMARY_CPU_MODEL,
        "physical_cpu_cores": PRIMARY_CPU_PHYSICAL_CORES,
        "logical_cpu_threads": PRIMARY_CPU_LOGICAL_THREADS,
        "ram_bytes": PRIMARY_RAM_BYTES,
        "cpu_governor": PRIMARY_CPU_GOVERNOR,
        "cpu_governor_policy_count": PRIMARY_CPU_GOVERNOR_POLICY_COUNT,
        "clock_synchronized": True,
    },
    "accelerator": {
        "gpu_count": PRIMARY_GPU_COUNT,
        "gpus": [
            {
                "index": 0,
                "name": PRIMARY_GPU_NAME,
                "compute_capability": PRIMARY_GPU_COMPUTE_CAPABILITY,
                "memory_total_mib": PRIMARY_GPU_MEMORY_MIB,
                "bf16_compute_supported": True,
            }
        ],
        "nvidia_driver_version": PRIMARY_NVIDIA_DRIVER_VERSION,
        "driver_cuda_api_version": PRIMARY_DRIVER_CUDA_API_VERSION,
        "cuda_driver_available": True,
        "host_cuda_toolkit_present": False,
        "primary_compute_dtype": "bf16",
        "persistence_mode": "Disabled",
        "compute_process_count": 0,
        "memory_used_mib": 0,
        "temperature_c": 40,
        "power_limit_w": 450.0,
        "application_graphics_clock_mhz": "[N/A]",
        "application_memory_clock_mhz": "[N/A]",
    },
}

CommandRunner = Callable[[Sequence[str]], str]
TextReader = Callable[[Path], str]
UnameProbe = Callable[[], platform.uname_result]
WhichProbe = Callable[[str], str | None]
GovernorProbe = Callable[[], Sequence[str]]


def _run_text(arguments: Sequence[str]) -> str:
    try:
        return subprocess.run(
            list(arguments),
            check=True,
            capture_output=True,
            text=True,
        ).stdout
    except (OSError, subprocess.CalledProcessError) as error:
        raise EnvironmentContractError(
            f"cannot probe primary host with {' '.join(arguments)!r}: {error}"
        ) from error


def _read_text(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except OSError as error:
        raise EnvironmentContractError(f"cannot read host fact {path}: {error}") from error


def _probe_governors() -> list[str]:
    paths = sorted(
        Path("/sys/devices/system/cpu/cpufreq").glob("policy*/scaling_governor"),
        key=lambda path: int(path.parent.name.removeprefix("policy")),
    )
    if not paths:
        raise EnvironmentContractError("host exposes no CPU frequency policies")
    return [_read_text(path).strip() for path in paths]


def _os_release(text: str) -> dict[str, str]:
    values: dict[str, str] = {}
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, raw_value = line.split("=", maxsplit=1)
        value = raw_value.strip()
        if len(value) >= 2 and value[0] == value[-1] and value[0] in {'"', "'"}:
            value = value[1:-1]
        values[key] = value
    return values


def _cpu_facts(text: str) -> tuple[str, int, int]:
    processors: list[dict[str, str]] = []
    for block in re.split(r"\n\s*\n", text.strip()):
        fields: dict[str, str] = {}
        for line in block.splitlines():
            if ":" in line:
                key, value = line.split(":", maxsplit=1)
                fields[key.strip()] = value.strip()
        if "processor" in fields:
            processors.append(fields)
    if not processors:
        raise EnvironmentContractError("/proc/cpuinfo contains no processor records")
    raw_models = {record.get("model name", "") for record in processors}
    if len(raw_models) != 1 or not next(iter(raw_models)):
        raise EnvironmentContractError("CPU model is missing or inconsistent across threads")
    raw_model = next(iter(raw_models))
    if "i7-13700K" in raw_model:
        cpu_model = "Intel Core i7-13700K"
    else:
        cpu_model = raw_model
    core_ids = {
        (record.get("physical id", "0"), record["core id"])
        for record in processors
        if "core id" in record
    }
    if not core_ids:
        raise EnvironmentContractError("/proc/cpuinfo lacks physical core identity")
    return cpu_model, len(core_ids), len(processors)


def _memory_bytes(text: str) -> int:
    match = re.search(r"(?m)^MemTotal:\s*([0-9]+)\s+kB\s*$", text)
    if match is None:
        raise EnvironmentContractError("/proc/meminfo lacks MemTotal in kB")
    return int(match.group(1)) * 1024


def _gpu_facts(query_text: str) -> list[dict[str, object]]:
    rows: list[dict[str, object]] = []
    for line_number, raw_line in enumerate(query_text.splitlines(), start=1):
        line = raw_line.strip()
        if not line:
            continue
        fields = [field.strip() for field in line.split(",")]
        if len(fields) != 11:
            raise EnvironmentContractError(
                f"nvidia-smi GPU row {line_number} has {len(fields)} fields, expected 11"
            )
        (
            index_text,
            name,
            driver,
            capability,
            memory_total_text,
            memory_used_text,
            temperature_text,
            persistence_mode,
            power_limit_text,
            graphics_clock_text,
            memory_clock_text,
        ) = fields
        try:
            index = int(index_text)
            memory_total_mib = int(memory_total_text)
            memory_used_mib = int(memory_used_text)
            temperature_c = int(temperature_text)
            power_limit_w = float(power_limit_text)
        except ValueError as error:
            raise EnvironmentContractError(
                f"nvidia-smi GPU row {line_number} has a non-numeric hardware fact"
            ) from error
        if not re.fullmatch(r"[0-9]+\.[0-9]+", capability):
            raise EnvironmentContractError(
                f"nvidia-smi GPU row {line_number} has invalid compute capability"
            )
        major = int(capability.split(".", maxsplit=1)[0])
        rows.append(
            {
                "index": index,
                "name": name,
                "driver_version": driver,
                "compute_capability": capability,
                "memory_total_mib": memory_total_mib,
                "bf16_compute_supported": major >= 8,
                "memory_used_mib": memory_used_mib,
                "temperature_c": temperature_c,
                "persistence_mode": persistence_mode,
                "power_limit_w": power_limit_w,
                "application_graphics_clock_mhz": graphics_clock_text,
                "application_memory_clock_mhz": memory_clock_text,
            }
        )
    if not rows:
        raise EnvironmentContractError("nvidia-smi reported no GPUs")
    if [row["index"] for row in rows] != list(range(len(rows))):
        raise EnvironmentContractError("nvidia-smi GPU indexes must be contiguous from zero")
    return rows


def _exact_equal(observed: object, expected: object, path: str) -> None:
    if type(observed) is not type(expected):
        raise EnvironmentContractError(
            f"{path}: expected type {type(expected).__name__}, got {type(observed).__name__}"
        )
    if isinstance(expected, dict):
        if set(observed) != set(expected):  # type: ignore[arg-type]
            observed_keys = set(observed)  # type: ignore[arg-type]
            raise EnvironmentContractError(
                f"{path}: keys differ; missing={sorted(set(expected) - observed_keys)}, "
                f"extra={sorted(observed_keys - set(expected))}"
            )
        for key, value in expected.items():
            _exact_equal(observed[key], value, f"{path}.{key}")  # type: ignore[index]
        return
    if isinstance(expected, list):
        if len(observed) != len(expected):  # type: ignore[arg-type]
            raise EnvironmentContractError(
                f"{path}: expected {len(expected)} items, got {len(observed)}"  # type: ignore[arg-type]
            )
        for index, (actual_item, expected_item) in enumerate(
            zip(observed, expected, strict=True)  # type: ignore[arg-type]
        ):
            _exact_equal(actual_item, expected_item, f"{path}[{index}]")
        return
    if observed != expected:
        raise EnvironmentContractError(
            f"{path}: expected {expected!r}, got {observed!r}"
        )


def _exact_keys(value: object, expected: set[str], path: str) -> dict[str, object]:
    if type(value) is not dict:
        raise EnvironmentContractError(f"{path}: expected object")
    result = value
    if set(result) != expected:
        raise EnvironmentContractError(
            f"{path}: keys differ; missing={sorted(expected - set(result))}, "
            f"extra={sorted(set(result) - expected)}"
        )
    return result


def _expect_int_range(value: object, path: str, minimum: int, maximum: int) -> int:
    if type(value) is not int or not minimum <= value <= maximum:
        raise EnvironmentContractError(
            f"{path}: expected integer in [{minimum}, {maximum}], got {value!r}"
        )
    return value


def _expect_positive_number(value: object, path: str) -> float:
    if type(value) not in {int, float} or not math.isfinite(float(value)) or value <= 0:
        raise EnvironmentContractError(f"{path}: expected positive finite number")
    return float(value)


def validate_environment_snapshot(value: object, path: str = "observed_environment") -> None:
    """Validate exact host identity plus bounded, fail-closed start conditions."""

    root = _exact_keys(
        value, {"schema_version", "environment_id", "host", "accelerator"}, path
    )
    _exact_equal(root["schema_version"], ENVIRONMENT_SNAPSHOT_SCHEMA_VERSION, f"{path}.schema_version")
    _exact_equal(root["environment_id"], PRIMARY_ENVIRONMENT_ID, f"{path}.environment_id")
    host = _exact_keys(
        root["host"],
        {
            "os_id",
            "os_version_id",
            "kernel_release",
            "machine",
            "cpu_model",
            "physical_cpu_cores",
            "logical_cpu_threads",
            "ram_bytes",
            "cpu_governor",
            "cpu_governor_policy_count",
            "clock_synchronized",
        },
        f"{path}.host",
    )
    expected_host = PRIMARY_ENVIRONMENT_SNAPSHOT["host"]
    if not isinstance(expected_host, dict):
        raise AssertionError("primary host contract must be an object")
    for key, expected in expected_host.items():
        _exact_equal(host[key], expected, f"{path}.host.{key}")

    accelerator = _exact_keys(
        root["accelerator"],
        {
            "gpu_count",
            "gpus",
            "nvidia_driver_version",
            "driver_cuda_api_version",
            "cuda_driver_available",
            "host_cuda_toolkit_present",
            "primary_compute_dtype",
            "persistence_mode",
            "compute_process_count",
            "memory_used_mib",
            "temperature_c",
            "power_limit_w",
            "application_graphics_clock_mhz",
            "application_memory_clock_mhz",
        },
        f"{path}.accelerator",
    )
    expected_accelerator = PRIMARY_ENVIRONMENT_SNAPSHOT["accelerator"]
    if not isinstance(expected_accelerator, dict):
        raise AssertionError("primary accelerator contract must be an object")
    for key in (
        "gpu_count",
        "gpus",
        "nvidia_driver_version",
        "driver_cuda_api_version",
        "cuda_driver_available",
        "host_cuda_toolkit_present",
        "primary_compute_dtype",
        "persistence_mode",
        "compute_process_count",
        "application_graphics_clock_mhz",
        "application_memory_clock_mhz",
    ):
        _exact_equal(
            accelerator[key], expected_accelerator[key], f"{path}.accelerator.{key}"
        )
    _expect_int_range(
        accelerator["memory_used_mib"], f"{path}.accelerator.memory_used_mib", 0, 256
    )
    _expect_int_range(
        accelerator["temperature_c"], f"{path}.accelerator.temperature_c", 0, 50
    )
    _expect_positive_number(
        accelerator["power_limit_w"], f"{path}.accelerator.power_limit_w"
    )


def environment_comparability_signature(value: object) -> dict[str, object]:
    """Return start conditions that must match between paired oracle producers."""

    validate_environment_snapshot(value)
    root = value
    accelerator = root["accelerator"]  # type: ignore[index]
    host = root["host"]  # type: ignore[index]
    return {
        "environment_id": root["environment_id"],  # type: ignore[index]
        "cpu_governor": host["cpu_governor"],  # type: ignore[index]
        "cpu_governor_policy_count": host["cpu_governor_policy_count"],  # type: ignore[index]
        "persistence_mode": accelerator["persistence_mode"],  # type: ignore[index]
        "power_limit_w": accelerator["power_limit_w"],  # type: ignore[index]
        "application_graphics_clock_mhz": accelerator["application_graphics_clock_mhz"],  # type: ignore[index]
        "application_memory_clock_mhz": accelerator["application_memory_clock_mhz"],  # type: ignore[index]
    }


def probe_primary_environment(
    *,
    command_runner: CommandRunner = _run_text,
    text_reader: TextReader = _read_text,
    uname_probe: UnameProbe = platform.uname,
    which_probe: WhichProbe = shutil.which,
    governor_probe: GovernorProbe = _probe_governors,
) -> dict[str, object]:
    """Collect and approve the actual primary host before canonical generation."""

    gpu_rows = _gpu_facts(
        command_runner(
            (
                "nvidia-smi",
                "--query-gpu=index,name,driver_version,compute_cap,memory.total,memory.used,temperature.gpu,persistence_mode,power.limit,clocks.applications.graphics,clocks.applications.memory",
                "--format=csv,noheader,nounits",
            )
        )
    )
    driver_versions = {str(row.pop("driver_version")) for row in gpu_rows}
    if len(driver_versions) != 1:
        raise EnvironmentContractError("GPUs report inconsistent NVIDIA drivers")
    smi_overview = command_runner(("nvidia-smi",))
    cuda_match = re.search(r"CUDA Version:\s*([0-9]+\.[0-9]+)", smi_overview)
    if cuda_match is None:
        raise EnvironmentContractError("nvidia-smi overview lacks driver CUDA API version")
    compute_processes = [
        line
        for line in command_runner(
            (
                "nvidia-smi",
                "--query-compute-apps=pid,process_name,used_memory",
                "--format=csv,noheader,nounits",
            )
        ).splitlines()
        if line.strip()
    ]
    governors = [value.strip() for value in governor_probe()]
    if not governors:
        raise EnvironmentContractError("host exposes no CPU frequency policies")
    clock_synchronized = command_runner(
        ("timedatectl", "show", "-p", "NTPSynchronized", "--value")
    ).strip()
    os_release = _os_release(text_reader(Path("/etc/os-release")))
    cpu_model, physical_cores, logical_threads = _cpu_facts(
        text_reader(Path("/proc/cpuinfo"))
    )
    ram_bytes = _memory_bytes(text_reader(Path("/proc/meminfo")))
    uname = uname_probe()
    snapshot: dict[str, object] = {
        "schema_version": ENVIRONMENT_SNAPSHOT_SCHEMA_VERSION,
        "environment_id": PRIMARY_ENVIRONMENT_ID,
        "host": {
            "os_id": os_release.get("ID", ""),
            "os_version_id": os_release.get("VERSION_ID", ""),
            "kernel_release": uname.release,
            "machine": uname.machine,
            "cpu_model": cpu_model,
            "physical_cpu_cores": physical_cores,
            "logical_cpu_threads": logical_threads,
            "ram_bytes": ram_bytes,
            "cpu_governor": (
                governors[0] if len(set(governors)) == 1 else "mixed"
            ),
            "cpu_governor_policy_count": len(governors),
            "clock_synchronized": clock_synchronized == "yes",
        },
        "accelerator": {
            "gpu_count": len(gpu_rows),
            "gpus": [
                {
                    key: row[key]
                    for key in (
                        "index",
                        "name",
                        "compute_capability",
                        "memory_total_mib",
                        "bf16_compute_supported",
                    )
                }
                for row in gpu_rows
            ],
            "nvidia_driver_version": next(iter(driver_versions)),
            "driver_cuda_api_version": cuda_match.group(1),
            "cuda_driver_available": True,
            "host_cuda_toolkit_present": which_probe("nvcc") is not None,
            "primary_compute_dtype": "bf16",
            "persistence_mode": gpu_rows[0]["persistence_mode"],
            "compute_process_count": len(compute_processes),
            "memory_used_mib": gpu_rows[0]["memory_used_mib"],
            "temperature_c": gpu_rows[0]["temperature_c"],
            "power_limit_w": gpu_rows[0]["power_limit_w"],
            "application_graphics_clock_mhz": gpu_rows[0][
                "application_graphics_clock_mhz"
            ],
            "application_memory_clock_mhz": gpu_rows[0][
                "application_memory_clock_mhz"
            ],
        },
    }
    validate_environment_snapshot(snapshot)
    return snapshot


EnvironmentProbe = Callable[[], dict[str, object]]

File: crates/riley-native/src/environment.rs
#![cfg_attr(all(test, not(feature = "cuda")), allow(dead_code))]

use std::collections::{BTreeMap, BTreeSet};
use std::error::Error;
use std::fmt;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use serde_json::{Value, json};

use crate::contract::PRIMARY_ENVIRONMENT_ID;

const EXPECTED_OS_ID: &str = "ubuntu";
const EXPECTED_OS_VERSION: &str = "22.04";
const EXPECTED_KERNEL: &str = "6.8.0-138-generic";
const EXPECTED_MACHINE: &str = "x86_64";
const EXPECTED_CPU: &str = "Intel Core i7-13700K";
const EXPECTED_PHYSICAL_CORES: usize = 16;
const EXPECTED_LOGICAL_THREADS: usize = 24;
const EXPECTED_RAM_BYTES: u64 = 67_185_598_464;
const EXPECTED_GOVERNOR: &str = "powersave";
const EXPECTED_POLICY_COUNT: usize = 24;
const EXPECTED_GPU_NAME: &str = "NVIDIA GeForce RTX 4090";
const EXPECTED_GPU_MEMORY_MIB: u64 = 24_564;
const EXPECTED_DRIVER: &str = "580.173.02";
const EXPECTED_DRIVER_CUDA_API: &str = "13.0";

#[derive(Clone, Debug, PartialEq)]
pub(crate) struct AcceleratorObservation {
    pub(crate) gpu_count: usize,
    pub(crate) index: u32,
    pub(crate) name: String,
    pub(crate) compute_capability: (u32, u32),
    pub(crate) memory_total_mib: u64,
    pub(crate) driver_version: String,
    pub(crate) driver_cuda_api_version: String,
    pub(crate) persistence_mode: String,
    pub(crate) compute_process_count: usize,
    pub(crate) memory_used_mib: u64,
    pub(crate) temperature_c: u32,
    pub(crate) power_limit_w: f64,
    pub(crate) application_graphics_clock_mhz: Option<u32>,
    pub(crate) application_memory_clock_mhz: Option<u32>,
}

#[derive(Debug)]
pub(crate) enum EnvironmentError {
    Io { path: PathBuf, source: io::Error },
    Invalid(String),
}

impl fmt::Display for EnvironmentError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io { path, source } => {
                write!(
                    formatter,
                    "cannot read environment fact {}: {source}",
                    path.display()
                )
            }
            Self::Invalid(reason) => write!(formatter, "primary environment differs: {reason}"),
        }
    }
}

impl Error for EnvironmentError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Io { source, .. } => Some(source),
            Self::Invalid(_) => None,
        }
    }
}

pub(crate) fn probe_primary_environment(
    accelerator: &AcceleratorObservation,
) -> Result<Value, EnvironmentError> {
    probe_primary_environment_from(
        Path::new("/etc/os-release"),
        Path::new("/proc/cpuinfo"),
        Path::new("/proc/meminfo"),
        Path::new("/proc/sys/kernel/osrelease"),
        Path::new("/sys/devices/system/cpu/cpufreq"),
        Path::new("/run/systemd/timesync/synchronized"),
        accelerator,
    )
}

#[allow(clippy::too_many_arguments, clippy::too_many_lines)]
fn probe_primary_environment_from(
    os_release_path: &Path,
    cpuinfo_path: &Path,
    meminfo_path: &Path,
    kernel_path: &Path,
    cpufreq_path: &Path,
    synchronized_path: &Path,
    accelerator: &AcceleratorObservation,
) -> Result<Value, EnvironmentError> {
    let os_release = parse_os_release(&read_text(os_release_path)?)?;
    let (cpu_model, physical_cores, logical_threads) = parse_cpuinfo(&read_text(cpuinfo_path)?)?;
    let ram_bytes = parse_meminfo(&read_text(meminfo_path)?)?;
    let kernel_release = read_text(kernel_path)?.trim().to_owned();
    let governors = read_governors(cpufreq_path)?;
    let clock_synchronized = synchronized_path.is_file();
    let host_cuda_toolkit_present = path_has_executable("nvcc");

    require_equal(
        "os_id",
        os_release.get("ID").map(String::as_str),
        EXPECTED_OS_ID,
    )?;
    require_equal(
        "os_version_id",
        os_release.get("VERSION_ID").map(String::as_str),
        EXPECTED_OS_VERSION,
    )?;
    require_equal("kernel_release", Some(&kernel_release), EXPECTED_KERNEL)?;
    require_equal("machine", Some(std::env::consts::ARCH), EXPECTED_MACHINE)?;
    require_equal("cpu_model", Some(&cpu_model), EXPECTED_CPU)?;
    require_number(
        "physical_cpu_cores",
        &physical_cores,
        &EXPECTED_PHYSICAL_CORES,
    )?;
    require_number(
        "logical_cpu_threads",
        &logical_threads,
        &EXPECTED_LOGICAL_THREADS,
    )?;
    require_number("ram_bytes", &ram_bytes, &EXPECTED_RAM_BYTES)?;
    if governors.len() != EXPECTED_POLICY_COUNT
        || governors
            .iter()
            .any(|governor| governor != EXPECTED_GOVERNOR)
    {
        return Err(EnvironmentError::Invalid(format!(
            "CPU governors must be {EXPECTED_POLICY_COUNT} copies of {EXPECTED_GOVERNOR:?}"
        )));
    }
    if !clock_synchronized {
        return Err(EnvironmentError::Invalid(
            "systemd clock synchronization marker is absent".to_owned(),
        ));
    }
    if host_cuda_toolkit_present {
        return Err(EnvironmentError::Invalid(
            "host CUDA toolkit must be absent from PATH".to_owned(),
        ));
    }
    validate_accelerator(accelerator)?;

    Ok(json!({
        "schema_version": "1.0.0",
        "environment_id": PRIMARY_ENVIRONMENT_ID,
        "host": {
            "os_id": EXPECTED_OS_ID,
            "os_version_id": EXPECTED_OS_VERSION,
            "kernel_release": kernel_release,
            "machine": EXPECTED_MACHINE,
            "cpu_model": cpu_model,
            "physical_cpu_cores": physical_cores,
            "logical_cpu_threads": logical_threads,
            "ram_bytes": ram_bytes,
            "cpu_governor": EXPECTED_GOVERNOR,
            "cpu_governor_policy_count": governors.len(),
            "clock_synchronized": true,
        },
        "accelerator": {
            "gpu_count": accelerator.gpu_count,
            "gpus": [{
                "index": accelerator.index,
                "name": accelerator.name,
                "compute_capability": format!("{}.{}", accelerator.compute_capability.0, accelerator.compute_capability.1),
                "memory_total_mib": accelerator.memory_total_mib,
                "bf16_compute_supported": accelerator.compute_capability.0 >= 8,
            }],
            "nvidia_driver_version": accelerator.driver_version,
            "driver_cuda_api_version": accelerator.driver_cuda_api_version,
            "cuda_driver_available": true,
            "host_cuda_toolkit_present": false,
            "primary_compute_dtype": "bf16",
            "persistence_mode": accelerator.persistence_mode,
            "compute_process_count": accelerator.compute_process_count,
            "memory_used_mib": accelerator.memory_used_mib,
            "temperature_c": accelerator.temperature_c,
            "power_limit_w": accelerator.power_limit_w,
            "application_graphics_clock_mhz": clock_text(accelerator.application_graphics_clock_mhz),
            "application_memory_clock_mhz": clock_text(accelerator.application_memory_clock_mhz),
        },
    }))
}

fn validate_accelerator(observed: &AcceleratorObservation) -> Result<(), EnvironmentError> {
    require_number("gpu_count", &observed.gpu_count, &1_usize)?;
    require_number("gpu.index", &observed.index, &0_u32)?;
    require_equal("gpu.name", Some(&observed.name), EXPECTED_GPU_NAME)?;
    if observed.compute_capability != (8, 9) {
        return Err(EnvironmentError::Invalid(format!(
            "compute capability is {:?}, expected (8, 9)",
            observed.compute_capability
        )));
    }
    require_number(
        "gpu.memory_total_mib",
        &observed.memory_total_mib,
        &EXPECTED_GPU_MEMORY_MIB,
    )?;
    require_equal(
        "nvidia_driver_version",
        Some(&observed.driver_version),
        EXPECTED_DRIVER,
    )?;
    require_equal(
        "driver_cuda_api_version",
        Some(&observed.driver_cuda_api_version),
        EXPECTED_DRIVER_CUDA_API,
    )?;
    require_equal(
        "persistence_mode",
        Some(&observed.persistence_mode),
        "Disabled",
    )?;
    require_number(
        "compute_process_count",
        &observed.compute_process_count,
        &0_usize,
    )?;
    if observed.memory_used_mib > 256 {
        return Err(EnvironmentError::Invalid(format!(
            "GPU memory used is {} MiB, maximum is 256",
            observed.memory_used_mib
        )));
    }
    if observed.temperature_c > 50 {
        return Err(EnvironmentError::Invalid(format!(
            "GPU temperature is {} C, maximum is 50",
            observed.temperature_c
        )));
    }
    if !observed.power_limit_w.is_finite() || (observed.power_limit_w - 450.0).abs() > f64::EPSILON
    {
        return Err(EnvironmentError::Invalid(format!(
            "GPU power limit is {}, expected 450 W",
            observed.power_limit_w
        )));
    }
    if observed.application_graphics_clock_mhz.is_some()
        || observed.application_memory_clock_mhz.is_some()
    {
        return Err(EnvironmentError::Invalid(
            "application clocks must both be unsupported ([N/A])".to_owned(),
        ));
    }
    Ok(())
}

fn read_text(path: &Path) -> Result<String, EnvironmentError> {
    fs::read_to_string(path).map_err(|source| EnvironmentError::Io {
        path: path.to_path_buf(),
        source,
    })
}

fn parse_os_release(text: &str) -> Result<BTreeMap<String, String>, EnvironmentError> {
    let mut values = BTreeMap::new();
    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let Some((key, raw_value)) = line.split_once('=') else {
            continue;
        };
        let raw_value = raw_value.trim();
        let value = if raw_value.len() >= 2
            && ((raw_value.starts_with('"') && raw_value.ends_with('"'))
                || (raw_value.starts_with('\'') && raw_value.ends_with('\'')))
        {
            &raw_value[1..raw_value.len() - 1]
        } else {
            raw_value
        };
        values.insert(key.to_owned(), value.to_owned());
    }
    if values.is_empty() {
        return Err(EnvironmentError::Invalid("os-release is empty".to_owned()));
    }
    Ok(values)
}

fn parse_cpuinfo(text: &str) -> Result<(String, usize, usize), EnvironmentError> {
    let mut models = BTreeSet::new();
    let mut cores = BTreeSet::new();
    let mut logical = 0_usize;
    for block in text.split("\n\n") {
        let fields: BTreeMap<_, _> = block
            .lines()
            .filter_map(|line| line.split_once(':'))
            .map(|(key, value)| (key.trim(), value.trim()))
            .collect();
        if fields.contains_key("processor") {
            logical += 1;
            if let Some(model) = fields.get("model name") {
                models.insert((*model).to_owned());
            }
            if let Some(core_id) = fields.get("core id") {
                cores.insert((fields.get("physical id").copied().unwrap_or("0"), *core_id));
            }
        }
    }
    if models.len() != 1 || cores.is_empty() || logical == 0 {
        return Err(EnvironmentError::Invalid(
            "CPU identity records are incomplete or inconsistent".to_owned(),
        ));
    }
    let raw_model = models.into_iter().next().unwrap_or_default();
    let model = if raw_model.contains("i7-13700K") {
        EXPECTED_CPU.to_owned()
    } else {
        raw_model
    };
    Ok((model, cores.len(), logical))
}

fn parse_meminfo(text: &str) -> Result<u64, EnvironmentError> {
    let line = text
        .lines()
        .find(|line| line.starts_with("MemTotal:"))
        .ok_or_else(|| EnvironmentError::Invalid("MemTotal is absent".to_owned()))?;
    let mut fields = line.split_ascii_whitespace();
    if fields.next() != Some("MemTotal:") {
        return Err(EnvironmentError::Invalid(
            "MemTotal is malformed".to_owned(),
        ));
    }
    let kib = fields
        .next()
        .ok_or_else(|| EnvironmentError::Invalid("MemTotal value is absent".to_owned()))?
        .parse::<u64>()
        .map_err(|_| EnvironmentError::Invalid("MemTotal is not numeric".to_owned()))?;
    if fields.next() != Some("kB") || fields.next().is_some() {
        return Err(EnvironmentError::Invalid(
            "MemTotal unit is not kB".to_owned(),
        ));
    }
    kib.checked_mul(1_024)
        .ok_or_else(|| EnvironmentError::Invalid("MemTotal overflows".to_owned()))
}

fn read_governors(root: &Path) -> Result<Vec<String>, EnvironmentError> {
    let entries = fs::read_dir(root).map_err(|source| EnvironmentError::Io {
        path: root.to_path_buf(),
        source,
    })?;
    let mut paths = entries
        .filter_map(Result::ok)
        .filter_map(|entry| {
            let name = entry.file_name();
            let name = name.to_str()?;
            let index = name.strip_prefix("policy")?.parse::<usize>().ok()?;
            Some((index, entry.path().join("scaling_governor")))
        })
        .collect::<Vec<_>>();
    paths.sort_unstable_by_key(|(index, _)| *index);
    if paths.is_empty() {
        return Err(EnvironmentError::Invalid(
            "no CPU frequency policies are visible".to_owned(),
        ));
    }
    paths
        .into_iter()
        .map(|(_, path)| Ok(read_text(&path)?.trim().to_owned()))
        .collect()
}

fn path_has_executable(name: &str) -> bool {
    std::env::var_os("PATH").is_some_and(|paths| {
        std::env::split_paths(&paths).any(|directory| directory.join(name).is_file())
    })
}

fn require_equal(
    name: &str,
    observed: Option<&str>,
    expected: &str,
) -> Result<(), EnvironmentError> {
    if observed == Some(expected) {
        Ok(())
    } else {
        Err(EnvironmentError::Invalid(format!(
            "{name} is {observed:?}, expected {expected:?}"
        )))
    }
}

fn require_number<T: fmt::Display + PartialEq>(
    name: &str,
    observed: &T,
    expected: &T,
) -> Result<(), EnvironmentError> {
    if observed == expected {
        Ok(())
    } else {
        Err(EnvironmentError::Invalid(format!(
            "{name} is {observed}, expected {expected}"
        )))
    }
}

fn clock_text(clock: Option<u32>) -> Value {
    clock.map_or_else(|| json!("[N/A]"), |value| json!(value.to_string()))
}

#[cfg(test)]
mod tests {
    use super::{
        AcceleratorObservation, parse_cpuinfo, parse_meminfo, parse_os_release,
        validate_accelerator,
    };

    fn accelerator() -> AcceleratorObservation {
        AcceleratorObservation {
            gpu_count: 1,
            index: 0,
            name: "NVIDIA GeForce RTX 4090".to_owned(),
            compute_capability: (8, 9),
            memory_total_mib: 24_564,
            driver_version: "580.173.02".to_owned(),
            driver_cuda_api_version: "13.0".to_owned(),
            persistence_mode: "Disabled".to_owned(),
            compute_process_count: 0,
            memory_used_mib: 200,
            temperature_c: 42,
            power_limit_w: 450.0,
            application_graphics_clock_mhz: None,
            application_memory_clock_mhz: None,
        }
    }

    #[test]
    fn parses_host_files_without_commands() {
        let os = parse_os_release("ID=ubuntu\nVERSION_ID=\"22.04\"\n").expect("os");
        assert_eq!(os["VERSION_ID"], "22.04");
        let cpu = parse_cpuinfo(concat!(
            "processor: 0\nphysical id: 0\ncore id: 0\nmodel name: 13th Gen Intel(R) Core(TM) i7-13700K\n\n",
            "processor: 1\nphysical id: 0\ncore id: 1\nmodel name: 13th Gen Intel(R) Core(TM) i7-13700K\n",
        ))
        .expect("cpu");
        assert_eq!(cpu, ("Intel Core i7-13700K".to_owned(), 2, 2));
        assert_eq!(
            parse_meminfo("MemTotal:       65500 kB\n").expect("mem"),
            67_072_000
        );
    }

    #[test]
    fn accelerator_contract_is_fail_closed() {
        validate_accelerator(&accelerator()).expect("valid");
        let mut changed = accelerator();
        changed.compute_process_count = 1;
        assert!(validate_accelerator(&changed).is_err());
        changed = accelerator();
        changed.power_limit_w = 449.0;
        assert!(validate_accelerator(&changed).is_err());
    }
}
